File: typescript/exercices/corrections/tp-final/services/productService.ts
import { Product } from "../classes/product"
import { BaseService } from "./baseService"

export class ProductService extends BaseService<Product> {
    constructor() {
        super("data/produits.json")    
    }

    addProduct(title:string, price:number, stock:number):void {
        const product = new Product(++this.count, title, price, stock)
        this.data.push(product)
        this.writeToJson()
    }

    getProductById(id:number):Product {
        return this.data.find(p => p.id == id)
    }

    updateProductStockById(id:number, deltaStock:number):boolean {
        const product:Product = this.getProductById(id)
        if(product != undefined) {
            product.stock -= deltaStock
            this.writeToJson()
            return true
        }
        return false
    }
}

File: FormationDotNet/CaisseEnregistreuse/Services/ProductService.cs
using CaisseEnregistreuse.Interfaces;
using CaisseEnregistreuse.Models;

namespace CaisseEnregistreuse.Services;
public class ProductService : IProductService
{
    private List<Product> _products;

    public ProductService()
    {
        _products = new List<Product>() { new Product() { Id =1, Title="product 1", Price=10, Stock=10} };
    }

    public List<Product> Products { get => _products; }

    public bool AddProduct(Product product)
    {
        product.Id = _products.Count + 1;
        _products.Add(product);
        return true;
    }

    public Product GetProductById(int id)
    {
        return _products.Find(p => p.Id == id);
    }
}

